libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -561,6 +554,71 @@ inline _CUDA_VSTD::uint64_t * barrier_native_handle(barrier<thread_scope_block>
+
+    _LIBCUDACXX_DEBUG_ASSERT(__isShared(barrier_native_handle(__b)), "Barrier must be located in local shared memory.");
+    _LIBCUDACXX_DEBUG_ASSERT(1 <= __arrive_count_update, "Arrival count update must be at least one.");
+    _LIBCUDACXX_DEBUG_ASSERT(__arrive_count_update <= (1 << 20) - 1, "Arrival count update cannot exceed 2^20 - 1.");
+    _LIBCUDACXX_DEBUG_ASSERT(__transaction_count_update >= 0, "Transaction count update must be non-negative.");
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#contents-of-the-mbarrier-object
+    _LIBCUDACXX_DEBUG_ASSERT(__transaction_count_update <= (1 << 20) - 1, "Transaction count update cannot exceed 2^20 - 1.");
+
+    barrier<thread_scope_block>::arrival_token __token = {};
+    NV_DISPATCH_TARGET(
+        NV_PROVIDES_SM_90, (
+            // We do not check for the statespace of the barrier here. This is
+            // on purpose. This allows debugging tools like memcheck/racecheck
+            // to detect that we are passing a pointer with the wrong state
+            // space to mbarrier.arrive. If we checked for the state space here,
+            // and __trap() if wrong, then those tools would not be able to help
+            // us in release builds. In debug builds, the error would be caught
+            // by the asserts at the top of this function.
+            if (__arrive_count_update == 1) {
+                asm volatile(
+                    "mbarrier.arrive.expect_tx.release.cta.shared::cta.b64 %0, [%1], %2;"
+                    : "=l"(__token)
+                    : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(barrier_native_handle(__b)))),
+                      "r"(static_cast<_CUDA_VSTD::uint32_t>(__transaction_count_update))
+                    : "memory");
+            } else {
+                asm volatile(
+                    "mbarrier.expect_tx.relaxed.cta.shared::cta.b64 [%0], %1;"
+                    :
+                    : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(barrier_native_handle(__b)))),
+                      "r"(static_cast<_CUDA_VSTD::uint32_t>(__transaction_count_update))
+                    : "memory");
+                asm volatile(
+                    "mbarrier.arrive.release.cta.shared::cta.b64 %0, [%1], %2;"
+                    : "=l"(__token)
+                    : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(barrier_native_handle(__b)))),
+                      "r"(static_cast<_CUDA_VSTD::uint32_t>(__arrive_count_update))
+                    : "memory");
+            }
+        ), NV_ANY_TARGET, (
+            // On architectures pre-SM90 (and in host code), arriving with a
+            // transaction count update is not supported.
+#if defined(__CUDA_MINIMUM_ARCH__) && __CUDA_MINIMUM_ARCH__ < 900
+            static_assert(
+                false,
+                "Insufficient CUDA Compute Capability: cuda::device::arrive_tx requires at least CC 9.0."
```

Comment:
This doesn't work, it'll fire when merely parsing this function with pre-SM90. This needs to be handled in some other way. (And `dependent_false` tricks also don't work, because there's no reason for this function to be a template.)

Reply:
Given that this is now a free function I see no reason why we would not guard it with the minimum arch so that it is not present when unavailable

Reply:
Oh right, it's testing minimum version, so that also works on NVC++, right?

Reply:
Done. the function is hidden on non-supported architectures.